components/style/properties/helpers/animated_properties.mako.rs

```diff
@@ -883,3 +883,69 @@ impl ToAnimatedZero for AnimatedFilter {
         }
     }
 }
+
+/// An iterator over all the properties that transition on a given style.
+pub struct TransitionPropertyIterator<'a> {
+    style: &'a ComputedValues,
+    index_range: core::ops::Range<usize>,
+    longhand_iterator: Option<NonCustomPropertyIterator<LonghandId>>,
+}
+
+impl<'a> TransitionPropertyIterator<'a> {
+    /// Create a `TransitionPropertyIterator` for the given style.
+    pub fn from_style(style: &'a ComputedValues) -> Self {
+        Self {
+            style,
+            index_range: 0..style.get_box().transition_property_count(),
+            longhand_iterator: None,
+        }
+    }
+}
+
+/// A single iteration of the TransitionPropertyIterator.
+pub struct TransitionPropertyIteration {
+    /// The id of the longhand for this property.
+    pub longhand_id: LonghandId,
+
+    /// The index of this property in the list of transition properties for this
+    /// iterator's style.
+    pub index: usize,
+}
+
+impl<'a> Iterator for TransitionPropertyIterator<'a> {
+    type Item = TransitionPropertyIteration;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        use crate::values::computed::TransitionProperty;
+
+        let next_longhand = self.longhand_iterator
+            .as_mut()
+            .and_then(|iterator| iterator.next());
+        if let Some(longhand_id) = next_longhand {
+            return Some(TransitionPropertyIteration {
+                longhand_id,
+                index: self.index_range.start,
+            });
+        }
+
+        let index = match self.index_range.next() {
+            Some(index) => index,
+            None => return None,
+        };
+
+        match self.style.get_box().transition_property_at(index) {
+            TransitionProperty::Custom(..) | TransitionProperty::Unsupported(..) => self.next(),
```

Comment:
May be better to make this a `loop { }` instead of recursive. I don't think it should be much harder.

Reply:
Sounds good. I've changed this too a loop and added a comment explaining what is going on.